tiledb/sm/rest/rest_client.cc

```diff
@@ -94,23 +95,42 @@ Status RestClient::get_array_schema_from_rest(
     const URI& uri, ArraySchema** array_schema) {
   STATS_FUNC_IN(rest_array_get_schema);
 
-  // Init curl and form the URL
-  Curl curlc;
-  RETURN_NOT_OK(curlc.init(config_, extra_headers_));
-  std::string array_ns, array_uri;
-  RETURN_NOT_OK(uri.get_rest_components(&array_ns, &array_uri));
-  std::string url = rest_server_ + "/v1/arrays/" + array_ns + "/" +
-                    curlc.url_escape(array_uri);
+  // Optimistically check for the cached array schema before
+  // locking.
+  if (serialized_array_schema_.data() == nullptr ||
+      serialized_array_schema_.size() == 0 ||
+      invalidate_serialized_array_schema_) {
+    serialized_array_schema_lock_.lock();
```

Comment:
This lock isn't being released on the error paths -- you can use a `std::lock_guard`, which will automatically release the lock when it goes out of scope.

Reply:
Ah, thanks. I did a quick scan for if-statements but neglected to consider the `RETURN_NOT_OK`.

Reply:
Err, scanned for `return` statements, not `if`.